### ISSUE TITLE
Authelia + Caddy v2 Configuration?

### ISSUE THREAD

**Issue by planecore** [authelia/authelia#1241]
Hello!
I noticed that Caddy v2 doesn't have a proxy integration on Authelia docs: https://www.authelia.com/docs/deployment/supported-proxies/

Are you planning on adding one?

**Comment by clems4ever** [comment]
Hello @planecore , it has been discussed with the core contributors but it's not really in our plans for now unfortunately.

**Comment by planecore** [comment]
Oh, I wish it'll be added sometime soon. I use Caddy 2 with [caddy-docker-proxy](https://github.com/lucaslorentz/caddy-docker-proxy) and it's much easier compared to Traefik.

**Comment by nightah** [comment]
@planecore to add to @clems4ever's comment this is primarily because when we last checked there did not appear to be a plugin which allows Caddy to allow/deny authentication based on the result of a subrequest.

Per my commentary [here](https://github.com/authelia/authelia/issues/1248#issuecomment-667768899), Authelia's integration with reverse proxies is based on [`ngx_http_auth_request_module`](http://nginx.org/en/docs/http/ngx_http_auth_request_module.html).
>The ngx_http_auth_request_module module (1.5.4+) implements client authorization based on the result of a subrequest. If the subrequest returns a 2xx response code, the access is allowed. If it returns 401 or 403, the access is denied with the corresponding error code. Any other response code returned by the subrequest is considered an error.

**Comment by akvadrako** [comment]
There is https://github.com/trusch/caddy-extauth which I think speaks the nginx protocol, though I suppose it's only useful as an example since it's not actively developed.

**Comment by james-d-elliott** [comment]
> 
> 
> There is https://github.com/trusch/caddy-extauth which I think speaks the nginx protocol, though I suppose it's only useful as an example since it's not actively developed.

Thanks for the link! Will see if I can get it working (or someone else is welcome to and ask for help).

**Comment by hslatman** [comment]
@james-d-elliott: this may be of use to you too: https://github.com/HeavenVolkoff/caddy-authelia. Haven't used it myself (yet), so I can't say if it works as expected, but it looks like a good start 🙂

**Comment by james-d-elliott** [comment]
Ah since the modules are available in go that makes it easy.

**Comment by hslatman** [comment]
I tried the module I linked in a custom built Caddy, but so far I couldn't get it to work correctly. In my setup I derived a basic configuration from the `local`/`lite` compose setups and created a Caddyfile according to the docs for the module. The first redirect to login works, but after logging in I'm redirected to the Authelia login again.

This is my first time running Authelia, so at this time I'm not sure if the issue I'm seeing is due to a misconfiguration of Authelia or due to the module not working correctly.

**Comment by james-d-elliott** [comment]
Has anyone tried this with the upstream plugin? https://github.com/freman/caddy-reauth#supported-backends

**Comment by james-d-elliott** [comment]
@hslatman do you know if the caddy-reauth plugin above works?

**Comment by Brendan-McDonnell** [comment]
@james-d-elliott It seems that plugin doesn't support Caddy v2, as [discussed here](https://github.com/freman/caddy-reauth/issues/32). There is a Caddy v2 version [here](https://github.com/freman/caddy2-reauth), but my impression is that it was a quick-and-dirty start and not something production ready (i.e. not even tested yet).

**Comment by james-d-elliott** [comment]
@Brendan-McDonnell thanks for the feedback, I will look at making a module to implement the same spec as traefik's forward auth I guess.

**Comment by james-d-elliott** [comment]
@Brendan-McDonnell can you see anything wrong with this? https://github.com/authelia/caddy-forwardauth/blob/v0.1.0/data/caddy/Caddyfile.json#L103-L124

Note that I'm not sure this works yet, just trying to test it however that reverse proxy handle seems to remove Content-Type headers.

**Comment by Brendan-McDonnell** [comment]
@james-d-elliott Two things:
1. Is there a reason this is being done via JSON instead of a Caddyfile? I believe most Caddy users use and are familiar with Caddyfiles instead of Caddy's underlying JSON system, especially for home/small projects. Unless we are using features not implemented in the Caddyfile or are programmatically generating the JSON, I would recommend Caddyfile instead just for its ubiquity among most Caddy users.
2. Headers should be passed through. There are controls for manipulating headers for the `reverse_proxy` directive, though, if necessary: [JSON docs](https://caddyserver.com/docs/json/apps/http/servers/routes/handle/reverse_proxy/headers/), [Caddyfile directive docs](https://caddyserver.com/docs/caddyfile/directives/reverse_proxy).

    The issue I'm seeing is that you are using an auth provider that does not exist. [These docs](https://caddyserver.com/docs/json/apps/http/servers/routes/handle/authentication/providers/) indicate that only basic auth is supported in Caddy's core reverse proxy directive, unfortunately. 

    [This issue](https://github.com/caddyserver/caddy/issues/2894) goes over the previous discussions on authentication systems with Caddy, including Authelia. The TL;DR is that there aren't currently plans to add more than basic auth into core Caddy given that [this plugin](https://github.com/greenpau/caddy-auth-portal) is active, significant, and covers most of the use cases. I would look into this plugin to see how Authelia can be integrated. It does have it's own Authelia-like features, so it may be better to see how they implement things on the backend instead if necessary.

    Unfortunately, I'm not sure if Authelia can be integrated into Caddy without a plugin supporting it, unless there is a push to advance the authentication means in base Caddy. If caddy-auth-portal doesn't work, it may be necessary to make some caddy-authelia bridging plugin...

**Comment by james-d-elliott** [comment]
>     1. Is there a reason this is being done via JSON instead of a Caddyfile? I believe most Caddy users use and are familiar with Caddyfiles instead of Caddy's underlying JSON system, especially for home/small projects. Unless we are using features not implemented in the Caddyfile or are programmatically generating the JSON, I would recommend Caddyfile instead just for its ubiquity among most Caddy users.

It's being done that way just while I test if it's theoretically possible. I'll look into the requirements of Caddyfile parsing and how to accomplish this once/if I can get this working for v1.0.0. It's just extra code.



>     2\. The issue I'm seeing is that you are using an auth provider that does not exist. [These docs](https://caddyserver.com/docs/json/apps/http/servers/routes/handle/authentication/providers/) indicate that only basic auth is supported in Caddy's core reverse proxy directive, unfortunately.

The way I've configured it should be working.. that docker-compose compiles caddy with the forwardauth authentication module. Looking at the reauth example this is exactly how it should be configured: https://github.com/freman/caddy2-reauth/blob/master/assets/conf/Caddyfile.json

Also the reverse proxy instance isn't using the authentication handle at all in my example unless I'm mistaken.

**Comment by Brendan-McDonnell** [comment]
Ah, I missed that a `forward_auth` plugin had been added. Oops!

`reverse_proxy` should pass all headers through with the exception of some specific hop-by-hop headers. What exactly is the expected/actual behavior of your testing there?

**Comment by akvadrako** [comment]
> [This issue](https://github.com/caddyserver/caddy/issues/2894) goes over the previous discussions on authentication systems with Caddy, including Authelia. The TL;DR is that there aren't currently plans to add more than basic auth into core Caddy given that [this plugin](https://github.com/greenpau/caddy-auth-portal) is active, significant, and covers most of the use cases. I would look into this plugin to see how Authelia can be integrated. It does have it's own Authelia-like features, so it may be better to see how they implement things on the backend instead if necessary.

I've done some evaluations of auth plugins for Caddy and I don't think using caddy-auth-portal with Authelia is an ideal setup. It only uses a proprietary cookie-based protocol between Caddy and it's builtin portal, which can then delegate to outside providers.

**Comment by james-d-elliott** [comment]
> 
> 
> > [This issue](https://github.com/caddyserver/caddy/issues/2894) goes over the previous discussions on authentication systems with Caddy, including Authelia. The TL;DR is that there aren't currently plans to add more than basic auth into core Caddy given that [this plugin](https://github.com/greenpau/caddy-auth-portal) is active, significant, and covers most of the use cases. I would look into this plugin to see how Authelia can be integrated. It does have it's own Authelia-like features, so it may be better to see how they implement things on the backend instead if necessary.
> 
> I've done some evaluations of auth plugins for Caddy and I don't think using caddy-auth-portal with Authelia is an ideal setup. It only uses a proprietary cookie-based protocol between Caddy and it's builtin portal, which can then delegate to outside providers.

I agree. I'd rather make a backend agnostic plugin that works similar to Traefik forward auth, which would make it work with Authelia by default.

**Comment by Brendan-McDonnell** [comment]
> > > [This issue](https://github.com/caddyserver/caddy/issues/2894) goes over the previous discussions on authentication systems with Caddy, including Authelia. The TL;DR is that there aren't currently plans to add more than basic auth into core Caddy given that [this plugin](https://github.com/greenpau/caddy-auth-portal) is active, significant, and covers most of the use cases. I would look into this plugin to see how Authelia can be integrated. It does have it's own Authelia-like features, so it may be better to see how they implement things on the backend instead if necessary.
> > 
> > 
> > I've done some evaluations of auth plugins for Caddy and I don't think using caddy-auth-portal with Authelia is an ideal setup. It only uses a proprietary cookie-based protocol between Caddy and it's builtin portal, which can then delegate to outside providers.
> 
> I agree. I'd rather make a backend agnostic plugin that works similar to Traefik forward auth, which would make it work with Authelia by default. I've done some initial development which anyone is welcome to help contribute to, however it's absolutely a side project of a side project for me at the moment.
> 
> See here: https://github.com/authelia/caddy-forwardauth/

Any further work on this? I would help contribute, but I have no experience in Golang.

Is there any testing I could do to help?

**Comment by xpufx** [comment]
As far as I understand caddy-forwardauth is not working yet. And caddy server team is not interested in adding this functionality to caddy itself because caddy-auth-portal (which is now caddy-security..before it was caddy-auth-jwt) exists?   caddy-security is a full idp implementation with a portal and everything. It's very complicated and does not have generic instructions to use pieces of it as far as I can see. It's a bit of a shame that in order to use an idp software properly one has to install a whole other idp suite as a module into the web server.

Is the caddy-forwardauth going to be a generic solution or need authelia?

Thanks for your work on this. I wish caddy folks would prioritize this.

**Comment by francislavoie** [comment]
> And caddy server team is not interested in adding this functionality to caddy itself because caddy-auth-portal (which is now caddy-security..before it was caddy-auth-jwt) exists?

Caddy team member here. I didn't even know this issue existed, nobody brought it to our attention. I wouldn't say we weren't interested, we were just never brought into the conversation!

I'm definitely interested in seeing if we could eventually bring in `caddy-forwardauth` as a standard module in Caddy, but we'll definitely need help. And if any interfaces need to be adjusted to make things easier, I'd be glad to help as well.

**Comment by mholt** [comment]
Caddy author here -- Francis pointed me to this. Let us know how we can help or if we can clarify anything. Would love to see some integration added here.

**Comment by akvadrako** [comment]
Great that the Caddy team is looking at this, but how is filing https://github.com/caddyserver/caddy/issues/2894 not bringing it to your attention?

**Comment by francislavoie** [comment]
@akvadrako I'm not sure what your goal is with your comment. What I said isn't untrue, we didn't know that there were open questions regarding Authelia.

That issue was opened to collect feedback on what adjustments could be made to Caddy's auth interface (which basicauth used) to make it a smoother experience for plugin authors. 

The discussion essentially derailed from that and didn't have a clear focus, and there wasn't any specific proposal for changes to Caddy. Nobody had written a proposal for actionable changes to Caddy. It was a lot of "would be nice if" and not so much "here's how it should look".

As Matt wrote, that specific issue had essentially run its course, and that new issues could be opened if an actionable proposal was written; that hasn't happened.

Either way, no need to dwell in the past. Let's move forwards.

**Comment by akvadrako** [comment]
> @akvadrako I think that 1 comment on the larger issue of 40 comments got a bit lost because the issue was practically resolved by the development of the https://github.com/greenpau/caddy-security "super-plugin" (really big plugin, ha).

It wasn't just that the comment was lost, but you explicitly rejected the proposal to support the extauth protocol. `caddy-security` is not an acceptable solution for the reasons mentioned above. It requires it's own server that forwards auth requests to the real authentication provider.

> I'm not very familiar with this space, but looking on our website, it seems that an Authelia plugin is already registered: https://caddyserver.com/docs/modules/http.handlers.authelia

The last comment about caddy-authelia was negative, though maybe things have changed. The repo still says it shouldn't be used in production. @hslatman reported:

> I tried the module I linked in a custom built Caddy, but so far I couldn't get it to work correctly

**Comment by mholt** [comment]
> It wasn't just that the comment was lost, but you explicitly rejected the proposal to support the extauth protocol.

Where? I don't remember doing that, and in a cursory search of the issue I can't find anywhere I mentioned extauth.

> caddy-security is not an acceptable solution for the reasons mentioned above. It requires it's own server that forwards auth requests to the real authentication provider.

Have you tried opening a feature request with that project? Paul is excellent to work with and very responsive in my experience.

**Comment by mholt** [comment]
@akvadrako Ah, I wasn't rejecting Authelia/extauth plugins in Caddy. I just said it was unlikely to be part of the standard distribution right now. New features generally start as external plugins first and then depending on demand, general utility, dependency, maintenance costs, and other factors, we can consider making it part of the standard distribution.

> I have and spent time trying to built a solution that avoided the middleman server. But he was not interested. It wouldn't really be a feature, but go against the whole design of his project.

Bummer. In that case, would one of the other extauth/Authelia plugins be an option to contribute to? Or make a new one that does what you need?

**Comment by akvadrako** [comment]
> Bummer. In that case, would one of the other extauth/Authelia plugins be an option to contribute to? Or make a new one that does what you need?

I'm not interested in maintaining a Caddy plugin or depending on a barely maintained one.

**Comment by james-d-elliott** [comment]
Hey @mholt and @francislavoie I'm so glad you reached out regarding this. I do want to get this working for caddy v2, in particular in a way that is *not* specific to Authelia and roughly replicates traefik's process as a means to make the integration headache for applications implementing it nearly zero.

I had done some work initially in replicating the traefik flow in what I expected to work but I ended up getting rather busy with other things. I would love to get some assistance in finishing this off. Is there a preferred method of contact for this purpose?

> I'm not very familiar with this space, but looking on our website, it seems that an Authelia plugin is already registered: https://caddyserver.com/docs/modules/http.handlers.authelia

This is actually not affiliated with Authelia for reference. I'd probably look at asking them to transfer it to us if we got it working, or picking another one entirely if they were unwilling.

> I'm not interested in maintaining a Caddy plugin or depending on a barely maintained one.

If we got the Caddy v2 plugin working at a level equivalent to traefik we'd implement integration tests with the latest version of Caddy similar to how we do integration tests with the latest version of nginx, haproxy, and traefik. In addition we'd ensure the dependencies are automatically upgraded via renovate. If Caddy ever wanted to merge it into their core or one of their official modules we'd be happy to donate the code and help with maintenance if necessary.

**Comment by francislavoie** [comment]
> I do want to get this working for caddy v2, in particular in a way that is not specific to Authelia and roughly replicates traefik's process as a means to make the integration headache for applications implementing it nearly zero.

I agree, I think following Traefik's ForwardAuth is probably the best approach that I can see; other auth providers seem to also support that method and it would be nice to piggy-back off that.

The other option is to follow Nginx's `auth_request` for similar reasons, but it's probably easier/better defined to go with Traefik's, also being Go.

> I would love to get some assistance in finishing this off. Is there a preferred method of contact for this purpose?

Here is fine, or an issue on the Caddy repo, or the Caddy forums, or @mholt could send you an invite to our Caddy Slack. Whatever you prefer.

> If Caddy ever wanted to merge it into their core or one of their official modules we'd be happy to donate the code and help with maintenance if necessary.

:+1: sounds good to me.

There's probably some code that can be shared between Caddy's `reverse_proxy` (e.g. `X-Forwarded-*` header trust, transport and TLS config).

**Comment by mholt** [comment]
Yeah, and to emphasize what we said ealier, we're very open to adding new functionality to the standard distribution -- but first we usually like to see that the plugin is very popular and generally useful before we just accept the responsibility and merge it in.

(Remember: in Caddy, everything is a plugin, even the HTTP server -- it's just a matter of deciding which ones we ship with our official distributions... which brings on liabilities, maintenance, bugs, etc.)

I'm in favor of having a more officially-supported Authelia/ForwardAuth plugin listed on the Caddy website too, so we can make that happen.

@james-d-elliott We'd like to invite you to our Caddy Slack (reserved for community helpers, maintainers, distribution team, sponsors, and plugin authors) -- let me know what email I can invite and we can help you with a tighter feedback loop there.

**Comment by james-d-elliott** [comment]
> @james-d-elliott We'd like to invite you to our Caddy Slack (reserved for community helpers, maintainers, distribution team, sponsors, and plugin authors) -- let me know what email I can invite and we can help you with a tighter feedback loop there.

Could you add `james.elliott` with the suffix being `authelia.com`, if that makes sense. Hopefully avoids bots adding me to a mail list!

**Comment by francislavoie** [comment]
We have an initial proof-of-concept here https://github.com/caddyserver/caddy/pull/4739!

I'm now getting @james-d-elliott's help with testing this idea, and if it seems to do the trick, I'll implement a `forward_auth` directive which is sugar for this `reverse_proxy` pattern.

**Comment by james-d-elliott** [comment]
Hey all, see this https://github.com/caddyserver/caddy/pull/4739#issuecomment-1118127824 for information on configuring caddy to work with Authelia.

Currently you need to build caddy in order to get this to work. Here is the Caddyfile, Dockerfile, and docker-compose I used.

```Caddyfile
auth.example.com {
	tls /etc/caddy/pub.chain.pem /etc/caddy/priv.pem
	log
	reverse_proxy authelia:9091
}

app1.example.com {
	tls /etc/caddy/pub.chain.pem /etc/caddy/priv.pem
	log
	route {
		reverse_proxy authelia:9091 {
			method GET
			rewrite "/api/verify?rd=https://auth.example.com"

			header_up X-Forwarded-Method {method}
			header_up X-Forwarded-Uri {uri}

			@good status 200
			@bad status 5xx

			@needs-auth {
				header Location *
				status 302 303 401
			}

			handle_response @good {
				request_header Remote-User {http.reverse_proxy.header.Remote-User}
				request_header Remote-Groups {http.reverse_proxy.header.Remote-Groups}
				# Add more headers to copy here
			}

			handle_response @needs-auth {
				respond "" {http.reverse_proxy.status_code}
				header Location {http.reverse_proxy.header.Location}
			}

			handle_response @bad {
				# copy some headers?
				error {http.reverse_proxy.status_text} {http.reverse_proxy.status_code}
			}
		}

		respond "Application: app1. User: {header.Remote-User}. Groups: {header.Remote-Groups}. Body: {http.request.body}."
	}
}

app2.example.com {
	tls /etc/caddy/pub.chain.pem /etc/caddy/priv.pem
	log
	route {
		reverse_proxy authelia:9091 {
			method GET
			rewrite "/api/verify?rd=https://auth.example.com"

			header_up X-Forwarded-Method {method}
			header_up X-Forwarded-Uri {uri}

			@good status 200
			@bad status 5xx

			@needs-auth {
				header Location *
				status 302 303 401
			}

			handle_response @good {
				request_header Remote-User {http.reverse_proxy.header.Remote-User}
				request_header Remote-Groups {http.reverse_proxy.header.Remote-Groups}
				# Add more headers to copy here
			}

			handle_response @needs-auth {
				respond {http.reverse_proxy.status_code}
				header Location {http.reverse_proxy.header.Location}
			}

			handle_response @bad {
				# copy some headers?
				error {http.reverse_proxy.status_text} {http.reverse_proxy.status_code}
			}
		}

		respond "Application: app1. User: {header.Remote-User}. Groups: {header.Remote-Groups}. Body: {http.request.body}."
	}
}
```

```Dockerfile
FROM caddy:2.5.0-builder AS builder

RUN xcaddy build proxy-pre-check

FROM caddy:2.5.0

COPY --from=builder /usr/bin/caddy /usr/bin/caddy
```

```yaml
version: "3.8"
services:
  caddy:
    container_name: caddy
    build: .
    restart: unless-stopped
    networks:
      caddy: {}
    ports:
      - 80:80
      - 443:443
    volumes:
      - ${PWD}/Caddyfile:/etc/caddy/Caddyfile
      - ${PWD}/pub.chain.pem:/etc/caddy/pub.chain.pem
      - ${PWD}/pub.pem:/etc/caddy/pub.pem
      - ${PWD}/priv.pem:/etc/caddy/priv.pem
  authelia:
    container_name: authelia
    image: ${AUTHELIA_REGISTRY:-docker.io}/${AUTHELIA_IMAGE:-authelia/authelia}${AUTHELIA_TAG:-:latest}
    restart: unless-stopped
    networks:
      caddy: {}
    expose:
      - 9091
    volumes:
      - ${PWD}/data/authelia/config:/config
networks:
  caddy:
    external: true
    name: caddy
```

**Comment by xpufx** [comment]
Apparently there's a new caddy-forward-auth module. I just saw it on caddy's download page. It's based on traefik's functionality. https://github.com/firecow/caddy-forward-auth

**Comment by francislavoie** [comment]
@xpufx the above config does the exact same as Traefik's ForwardAuth, but without any plugins.

**Comment by nightah** [comment]
So I've just had integration tests pass with the latest commit which resolved the `X-Forwarded-Host` header missing the port.
Tested locally and the following build will represent said changes: https://buildkite.com/authelia/authelia/builds/11335.

Got some updating to do to our docs/README etc, we will probably hold off on suggesting Caddy is supported until these changes make it into an actual upstream release.

@francislavoie I know you mentioned earlier that you may consider a `forward_auth` directive, I think one of the main things to consider in the said directive is the ability to provide what headers should be passed across as options on a successful/authenticated request. Other than that things are looking good!

**Comment by francislavoie** [comment]
@nightah yeah, that's the main thing that the sugar directive would make easier, it would just take a list of headers to copy over. I'll work on that soon.

**Comment by francislavoie** [comment]
The `forward_auth` subdirective is now implemented! It looks like this:

```
auth.example.com {
	reverse_proxy authelia:9091
}

app1.example.com {
	forward_auth authelia:9091 {
		uri /api/verify?rd=https://auth.example.com
		copy_headers Remote-User Remote-Groups Remote-Name Remote-Email
	}

	respond "Application: app1. User: {header.Remote-User}. Groups: {header.Remote-Groups}. Body: {http.request.body}."
}
```

https://github.com/caddyserver/caddy/pull/4739 will be merged soon, and this'll land in v2.5.1 :tada:

**Comment by francislavoie** [comment]
Caddy v2.5.1 is released https://github.com/caddyserver/caddy/releases/tag/v2.5.1

Docs for the new `forward_auth` directive will come soon.
